tests/configlet/util/base_test.py

```diff
@@ -0,0 +1,301 @@
+
+
+# Test Description:
+# configlet is a generic tool that can patch any entry in CONFIG-DB.
+# AddRack is one of the scenarios that uses configlet
+#
+# AddRack:
+#   This is part of building of a new cluster with a T0. 
+#   Using AddRack-configlet, this new T0 could be added to a T1
+#   This includes the following
+#
+#   1) Use CLI (config ...) to ensure that this port is marked as down
+#
+#   Creates a configlet per strict template that is published in OneNote.
+#   (a) Adds a new Device Neighbor
+#   (b) Adds a new Portchannel for this port
+#   (c) Appends to ACL_TABLE for Everflow
+#   (d) Adds other port related entries 
+#       CABLE_LENGTH, QUEUE, BUGGER_PG, BUFFER_QUEUE, BUFFER_PG,
+#       BUFFER_PORT_INGRESS_PROFILE_LIST & BUFFER_PORT_EGRESS_PROFILE_LIST, if 
+#       Mellanox, PORT_QOS_MAP, PFC_WD
+#   (e) Add a new BGP Neighbor
+#   
+#   2) Apply the above template using "configlet" tool
+#   3) Use CLI to set this port up
+# 
+# How this test works ?
+#   (a) Load the current/original minigraph for this device, which has all T0s
+#      per its topology
+#   (b) Take a dump of CONFIG-DB, APP-DB & ASIC-DB
+#   (c) Generate a new minigraph, by removing a T0 from the original minigraph
+#   (d) Load the new minigraph
+#   (e) Generate the configlet for this removed T0
+#   (f) Apply this configlet
+#   (g) Get dump of CONFIG-DB, APP-DB & ASIC-DB
+#   (h) Verify the following
+#       (a) BGP is up for this new neighbor
+#       (b) Compare the DB dumps with the dumps taken with original minigraph
+#
+# Upon successful testing via configlet using strict template, retry the 
+# same using generic updater.
+# During above test, it takes a copy of config_db.json w/o T0 
+# Takes a diff between config_db.json original & w/o T0.
+# Apply this patch using CLI (config apply-patch ...) which uses generic
+# updater. 
+#   1) Load config_db.json w/o T0
+#   2) Create patch from config_db.json w/o To to original
+#   3) Apply the patch
+#   4) Take DB dumps & compare with original
+#   5) Check BGP session
+#
+# Upon successful testing of patch application using generic updater,
+# test removal of same T0
+#
+#   1) Create patch from config_db.json original to w/o T0
+#   2) Apply patch (confilg apply-patch ...) 
+#   3) Take DB dumps and compare with those taken upon loading minigraph w/o T0
+#
+
+def init(duthost):
+    global init_data
+
+    if not os.path.exists(base_dir):
+        os.system("mkdir -p {}".format(base_dir))
+
+    for i in [ data_dir, orig_db_dir, no_t0_db_dir, clet_db_dir,
+            patch_add_t0_dir, patch_rm_t0_dir, files_dir ]:
+        os.mkdir(i)
+
+    init_data["files_dir"] = files_dir 
+    init_data["data_dir"] = data_dir
+    init_data["orig_db_dir"] = orig_db_dir
+    init_data["switch_name"] = duthost.hostname
+
+    init_data["version"] = duthost.os_version
+    log_debug("Created data_dir={} version={}".format(data_dir, init_data["version"]))
+
+
+def backup_minigraph(duthost):
```

Comment:
>backup_minigraph [](http://example.com/codeflow?start=4&length=16)

Is it better to leverage the checkpoint feature in GCU? It will backup/restore ConfigDB.

Reply:
This does in special way of creating a copy as /etc/sonic/orig/minigraph.xml.addRack.orig. I prefer it that way.
This feature is exposed to both manual & Ansible.